commcare_connect/opportunity/visit_import.py

```diff
@@ -311,3 +313,173 @@ def get_status_by_completed_work_id(dataset):
+        try:
+            longitude = Decimal(self.row[index])
+        except (ValueError, TypeError, InvalidOperation):
+            raise InvalidValueError(f"Invalid longitude value: {self.row[index]}")
+        if not Decimal("-180") <= longitude <= Decimal("180"):
+            raise InvalidValueError("Longitude must be between -180 and 180 degrees")
+        return longitude
+
+    def _get_radius(self) -> int:
+        index = _get_header_index(self.headers, RADIUS_COL)
+        try:
+            radius = int(self.row[index])
+        except (ValueError, TypeError):
+            raise InvalidValueError(f"Invalid radius value: {self.row[index]}")
+        if radius <= 0:
+            raise InvalidValueError("Radius must be a positive integer")
+        return radius
+
+    def _get_active(self) -> bool:
+        error_message = "Active status must be 'yes' or 'no'"
+        index = _get_header_index(self.headers, ACTIVE_COL)
+        active = self.row[index]
+        if not active:
+            raise InvalidValueError(error_message)
+        active = active.lower().strip()
+        if active not in ["yes", "no"]:
+            raise InvalidValueError(error_message)
+        return active == "yes"
+
+    def _get_area_name(self) -> str:
+        index = _get_header_index(self.headers, AREA_NAME_COL)
+        area_name = self.row[index]
+        if not area_name:
+            raise InvalidValueError("Area name is not valid.")
+        return area_name
+
+    def _get_username(self) -> str | None:
+        try:
+            index = _get_header_index(self.headers, USERNAME_COL)
+        except ImportException:
+            return None
+        username = self.row[index]
+        return username if username else None
+
+    def _get_id(self) -> int | None:
+        try:
+            index = _get_header_index(self.headers, ID)
+        except ImportException:
+            return None
+        catchment_id = self.row[index]
+        return catchment_id if catchment_id else None
+
+
+def create_or_update_catchment(row_data: RowData, opportunity: Opportunity, opportunity_access: dict):
+    if row_data.id:
+        catchment = None
+        try:
+            catchment = CatchmentArea.objects.get(id=row_data.id)
+        except CatchmentArea.DoesNotExist:
+            None
+        if catchment:
+            catchment.latitude = row_data.latitude
+            catchment.longitude = row_data.longitude
+            catchment.radius = row_data.radius
+            catchment.name = row_data.area_name
+            catchment.active = row_data.active
+            return catchment, False
+
+    return (
+        CatchmentArea(
+            latitude=row_data.latitude,
+            longitude=row_data.longitude,
+            radius=row_data.radius,
+            opportunity=opportunity,
+            name=row_data.area_name,
+            active=row_data.active,
+            opportunity_access=opportunity_access.get(row_data.username, None),
```

Comment:
This is quite confusing because the variable `opportunity_access` isn't an opportunity access model, but the field `opportunity_access` is. It would be nice to rename the dict as something like `username_to_oa_map` or something similar that is a bit more descriptive.

Reply:
Changed